[cli] Added auth commands to expo

EvanBacon opened the pull request:
# Why

Authentication is required for signing manifests in `expo start`, otherwise you cannot use secure features in Expo Go. Since user info is used for `expo start` manifest middleware, authentication might as well be here as well. 

The API contents were copied over from `expo/eas-cli` which were also copied from `expo/expo-cli`, I've opted to use the newer API since the older one is unmaintained.

The newer API appears to implement a multitude of heavy packages which differ from what we used originally in expo-cli. Will need to take a pass at rewriting features using got to use node-fetch (`eas-cli` uses both).

- Removed amplitude code, renamed analytics to telemetry, removed persistent process settings.

# Test Plan

- Added a series of e2e tests for the auth commands.
- Migrated `eas-cli` tests
- Added extra tests and reduced unneeded things from eas-cli.
- Ran `nexpo login`, `nexpo logout`, `nexpo register`, `nexpo whoami` locally to ensure everything works.

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

EvanBacon committed:
Added auth commands to expo/expo

EvanBacon committed:
Added api tests

Update otp.ts

Update actions.ts

EvanBacon committed:
Update generated.ts

EvanBacon committed:
Update index-test.ts

EvanBacon committed:
Merge branch 'main' into @evanbacon/expo/cli/auth-commands

EvanBacon committed:
Added module loading tests

drop amplitude

Update rudderstackClient.ts

ide commented on packages/expo/cli/login/index.ts:
```diff
@@ -0,0 +1,51 @@
+    Log.exit(
+      chalk`
+      {bold Description}
+        Login to an Expo account
```
```suggestion
        Log in to an Expo account
```

ide commented on packages/expo/cli/logout/index.ts:
```diff
@@ -0,0 +1,38 @@
+    Log.exit(
+      chalk`
+      {bold Description}
+        Logout of an Expo account
```
```suggestion
        Log out of an Expo account
```

ide commented on packages/expo/cli/register/registerAsync.ts:
```diff
@@ -0,0 +1,28 @@
+import { ora } from '../utils/ora';
+
+export async function registerAsync() {
+  const REGISTRATION_URL = `https://expo.dev/signup`;
+  if (CI) {
+    throw new CommandError(
+      'NON_INTERACTIVE',
+      `Cannot register an account in CI. Register an account at: ${REGISTRATION_URL}`
+    );
+  }
```
IMO it would be better to suggest creating a bot user for your organization or using a personal access token, documented here: https://docs.expo.dev/accounts/programmatic-access/

byCedric replied:
I think matching [this](https://github.com/expo/expo/pull/16087/files#diff-21fde67c59cfa3d31ff4332a0c53e83b7b32cd425d12bec562a2a100a4b2ae2fR49-R51) would be great.
```js
{
  nonInteractiveHelp: `Use the EXPO_TOKEN environment variable to authenticate in CI (${learnMore(
    'https://docs.expo.dev/accounts/programmatic-access/'
  )})`,
 }
```

ide commented on packages/expo/cli/utils/analytics/rudderstackClient.ts:
```diff
@@ -0,0 +1,112 @@
+import RudderAnalytics from '@expo/rudder-sdk-node';
```
@kgc00 can you review this part of the PR?

kgc00 replied:
Thanks for roping me in James, and good work on the PR Evan! Unfortunately, this would disrupt our current analytics in a few ways. Here are some problems I saw:
- we're using eas cli's write keys so events would end up in the wrong place
- we're now using a single client. Previously, expo-cli used two clients: a legacy XDL client and  a newer 'Unified' client. If we don't adopt that methodology events will also end up in the place.
- we're introducing a staging/prod distinction to events where there previously was only a prod pipeline. Probably a good idea in the long run but needs some backend setup and consideration if it's proper to do now.
- we're drops all tracking of events (we have the rudder client, but it is not used anywhere)

I also think it would be good to retain some common logic inside of a `Command` or `ExpoCommand` that all other commands subclass. This way we do not need to rely on developers remembering to add in track calls or initializing clients in the proper place when adding new commands.

I think it's a really good start, and I'm just thinking about how to address these issues. Will jump back in with suggestions later today when I have some answers.

EvanBacon replied:
This is copied from eas-cli and reworked a bit to remove unused functions, structurally it's the same.

kgc00 replied:
hey! before i reply to some of the other comments:
> This is copied from eas-cli and reworked a bit to remove unused functions, structurally it's the same.

I have the feeling I am misunderstanding something, so sorry for the confusion. These analytics commands are expo-cli's commands, correct? e.g. this PR is not introducing eas-cli commands?

EvanBacon replied:
@kgc00 both expo-cli and eas-cli implement identical commands for authentication, these copy the code from the newer eas-cli instead of the older expo-cli.

kgc00 replied:
> @kgc00 both expo-cli and eas-cli implement identical commands for authentication, these copy the code from the newer eas-cli instead of the older expo-cli.

cool, so based on your response I'm going to assume this PR is specifically for expo-cli events. In that case, I want to explain some differences between how analytics are set up between Expo CLI and EAS analytics:
- they use different write keys and end up in different places in our backends
  - Additionally, Expo CLI has two different clients which write to two different places in our backend
- Expo CLI has no distinction between staging and prod events

Just copying over how things are done in EAS will break existing pipelines (which I tried to point out in my first comments, sorry if that wasn't clear). I did speak with the folks using this data yesterday, so I will add in some suggestions shortly to fix these issues. We will be replicating the setup expo cli was using so we don't break event streams.

> I'd rather we import the analytics explicitly for every command we want to track rather than wrapping every command with a catch-all that could be (and based on expo-cli and eas-cli will be) bloated very quickly.

I think we might disagree here. The analytics in expo-cli were broken for who knows how long because they were initialized differently for each command. This led to trying to track events with no client available. I also think the eas entry point is quite clean, it does 2 things. The expo entry point is quite old, so I don't think it should be used as an example of best practices. I also think the additional time-to-start is worth the reduced complexity for having some location which can share global setup/state. But that's my two cents and I don't want to block you on this PR.

 > we don't track any events yet there is no start command in the CLI yet, please check usage
 
 Cool, thanks for clearing that up. Mind adding me in as a reviewer for future PRs touching analytics? thanks!

kgc00 replied:
spoke with Evan and some others outside of the pr, we'll be going with a lighter-weight approach (reflected in the updated suggestions).

ide commented on packages/expo/cli/utils/graphql/client.ts:
```diff
@@ -0,0 +1,93 @@
+    throw error;
+  }
+
+  // Check for malfolmed response. This only checks the root query existence,
+  // It doesn't affect returning responses with empty resultset.
```
Let's clean this up while we're at it.
```suggestion
  // Check for a malformed response. This only checks the root query's existence. It doesn't affect
  // returning responses with an empty result set.
```

ide commented on packages/expo/cli/utils/api.ts:
```diff
@@ -0,0 +1,95 @@
+  }
+}
+
+export function getEASUpdateURL(projectId: string): string {
+  if (EXPO_STAGING) {
+    return new URL(projectId, `https://staging-u.expo.dev`).href;
+  } else if (EXPO_LOCAL) {
+    return new URL(`expo-updates/${projectId}`, `http://127.0.0.1:3000`).href;
+  } else {
+    return new URL(projectId, `https://u.expo.dev`).href;
+  }
+}
```
Probably can remove this since code that communicates with EAS Update will live in EAS CLI.

ide commented on packages/expo/cli/utils/graphql/client.ts:
```diff
@@ -0,0 +1,93 @@
+  'expo-session': string;
+};
+
+export const graphqlClient = createUrqlClient({
```
I know this is orthogonal to this PR but I would like to fix/get rid of that CombinedError message later unless there truly are multiple errors, in which case there is still probably a nicer way to print them than with [CombinedError]

ide commented on packages/expo/cli/utils/user/otp.ts:
```diff
@@ -0,0 +1,179 @@
+import assert from 'assert';
```
@wschurman Could you look over the login/auth-related files? I believe they are taken straight from the other CLIs for the most part so just need a sanity check

wschurman replied:
LGTM

byCedric commented on packages/expo/cli/utils/user/actions.ts:
```diff
@@ -0,0 +1,99 @@
+import { retryUsernamePasswordAuthWithOTPAsync } from './otp';
+import { Actor, getUserAsync, loginAsync } from './user';
+
+/** Show login prompt while prompting for missing credentials. */
```
Anyone: _writes comment_
James:
```suggestion
/** Show log in prompt while prompting for missing credentials. */
```

byCedric replied:
Sorry, couldn't let this one slip 😂

kbrandwijk replied:
It's actually 'login prompt'. The verb is 'to log in' (separate words), but the noun is 'login'.

ide replied:
Kim is right. "The setup steps", "Set up your project"

byCedric replied:
jeez

![its-a-meme](https://user-images.githubusercontent.com/1203991/151408258-b046e135-5a5d-4df1-9c30-8dd1b39ce7b5.gif)

byCedric commented on packages/expo/cli/utils/user/actions.ts:
```diff
@@ -0,0 +1,99 @@
+
+/** Ensure the user is logged in, if not, prompt to login. */
+export async function ensureLoggedInAsync(): Promise<Actor> {
+  let user: Actor | undefined;
+  try {
+    user = await getUserAsync();
+  } catch {}
```
Random question, but what do you think about this?
```js
let user = await getUserAsync().catch(() => undefined);
```

kbrandwijk replied:
I like `await...catch(...)` better, the Promise already comes out of the box with a catch mechanism, no need to wrap it in a separate one.

byCedric reviewed (commented):
Looking good! Left a few things, mainly the `if (EXPO_NO_TELEMETRY) // enable rudderstack`

ide commented on packages/expo/cli/utils/user/__tests__/otp-test.ts:
```diff
@@ -0,0 +1,256 @@
+jest.mock('../user');
+jest.mock('../../../log');
+
+const asMock = (fn: any): jest.Mock => fn as jest.Mock;
```
That's nice. We should do that in www more too. Will take a look today.

byCedric replied:
I used it in the expo-github-action update, if you want to try it out: https://github.com/expo/expo-github-action/blob/main/tests/expo.test.ts#L18

EvanBacon replied:
we'd need to update jest across the repo, that's out of scope for this PR.

EvanBacon commented on packages/expo/e2e/__tests__/whoami-test.ts:
```diff
@@ -0,0 +1,79 @@
+  }
+});
+
+if (process.env.CI) {
```
The test above this one runs in any environment and tests against being authenticate or unauthenticated. This CI test will only test against unauthenticated environments.

EvanBacon commented on packages/expo/cli/utils/user/actions.ts:
```diff
@@ -0,0 +1,99 @@
+    Log.log();
+  }
+
+  Log.log(hasCredentials ? 'Logging in to EAS' : 'Log in to EAS');
```
Seems like we should pick one or the other for all tools

kbrandwijk commented on packages/expo/cli/register/registerAsync.ts:
```diff
@@ -0,0 +1,28 @@
+import { ora } from '../utils/ora';
+
+export async function registerAsync() {
+  const REGISTRATION_URL = `https://expo.dev/signup`;
```
Should we use `getExpoWebsiteBaseUrl()` here?

EvanBacon replied:
That method is used for cases when the URL changes to test staging or dev environments. This URL doesn't really benefit from changing.

EvanBacon replied:
this command just opens a link, I think the link could be opened manually when testing, it seems over complicated.

kbrandwijk replied:
```js
const REGISTRATION_URL = `${getExpoWebsiteBaseUrl()}/signup`;
```
...or similar, isn't too overly complicated with the benefit of consistency, right? It's not like we need to build an entire thing around it.

EvanBacon replied:
Importing `getExpoApiBaseUrl` also imports a new file, the file it's in increases the JS bin size by a fair amount.
So we split it out into a new file, but now we have this weird doc block in the file saying we need the file to only contain one function because we use it in the `expo register` command for 'consistency reasons'.

kbrandwijk commented on packages/expo/cli/utils/user/UserSettings.ts:
```diff
@@ -0,0 +1,40 @@
+// Extracted from https://github.com/sindresorhus/env-paths/blob/main/index.js
+function getConfigDirectory() {
+  // Share data between eas-cli and expo.
+  const name = 'eas-cli';
```
Is this still correct here?

EvanBacon replied:
ya see comment -- sharing means the settings will be the same across tools.

kbrandwijk commented on packages/expo/e2e/__tests__/logout-test.ts:
```diff
@@ -0,0 +1,55 @@
+
+it('runs `npx expo logout --help`', async () => {
+  const results = await execute('logout', '--help');
+  expect(results.stdout).toMatchInlineSnapshot(`
```
I remember some kind of clean up function from another PR to remove the special characters. removeAnsiSomething...? Use here for consistency? (And in other similar test above and below)

kbrandwijk replied:
https://github.com/expo/expo/blob/main/packages/expo/cli/utils/__tests__/utils.ts#L1

EvanBacon committed:
Apply suggestions from code review

Co-authored-by: James Ide <ide@users.noreply.github.com>
Co-authored-by: Cedric van Putten <me@bycedric.com>

EvanBacon committed:
clean up

lazy load analytics

Update prompts.ts

Update actions.ts

EvanBacon commented on packages/expo/cli/utils/api.ts:
```diff
@@ -0,0 +1,75 @@
+import { JSONValue } from '@expo/json-file';
+import got, { HTTPError, NormalizedOptions, RequestError } from 'got';
```
My understanding was that we agreed not to use axios or got so we'll need to circle back and replace this with node-fetch.

EvanBacon committed:
fix tests

EvanBacon committed:
Update user.ts

kgc00 commented on packages/expo/cli/utils/analytics/rudderstackClient.ts:
```diff
@@ -0,0 +1,99 @@
+  linux: 'Linux',
+};
+
+let client: RudderAnalytics | null = null;
```
```suggestion
let unifiedClient: RudderAnalytics | null = null;
let xdlClient: RudderAnalytics | null = null;
```

kgc00 commented on packages/expo/cli/utils/analytics/rudderstackClient.ts:
```diff
@@ -0,0 +1,99 @@
+  traits: Record<string, any>;
+} | null = null;
+
+function getClient(): RudderAnalytics {
+  if (client) {
+    return client;
+  }
+
+  client = new RudderAnalytics(
+    EXPO_STAGING || EXPO_LOCAL ? '1wpX20Da4ltFGSXbPFYUL00Chb7' : '1wpXLFxmujq86etH6G6cc90hPcC',
+    'https://cdp.expo.dev/v1/batch',
+    {
+      flushInterval: 300,
+    }
+  );
+
+  // Install flush on exit...
+  process.on('SIGINT', () => client?.flush?.());
+  process.on('SIGTERM', () => client?.flush?.());
+
+  return client;
+}
```
```suggestion
function _getUnifiedClient(): RudderAnalytics {
  if (unifiedClient) {
    return unifiedClient;
  }

  unifiedClient = new RudderAnalytics(
     '1wabJGd5IiuF9Q8SGlcI90v8WTs',
    'https://cdp.expo.dev/v1/batch',
    {
      flushInterval: 300,
    }
  );

  // Install flush on exit...
  process.on('SIGINT', () => client?.flush());
  process.on('SIGTERM', () => client?.flush());

  return client;
}
```

kgc00 replied:
also, it would be nice to have some async context where we can `await client.flush` on exit.

kgc00 commented on packages/expo/cli/utils/analytics/rudderstackClient.ts:
```diff
@@ -0,0 +1,99 @@
+  deviceId: string;
+  traits: Record<string, any>;
+} | null = null;
+
```
should wrap all these methods in an Analytics class like in XDL https://github.com/expo/expo-cli/blob/main/packages/xdl/src/Analytics.ts

e.g. create an internal analytics class here, then export two versions of it with different write keys and have the caller specify which client they need

```
class AnalyticsClient {
  private userTraits?: { [key: string]: any };
  private rudderstackClient?: RudderAnalytics;
  private _userId?: string;
  private _version?: string;
  ....
}

const xdlClient = new AnalyticsClient();
const unifiedClient = new AnalyticsClient();

export { xdlClient, unifiedClient };
```

kgc00 commented on packages/expo/cli/utils/analytics/rudderstackClient.ts:
```diff
@@ -0,0 +1,99 @@
+  linux: 'Linux',
+};
+
+let client: RudderAnalytics | null = null;
+let identified = false;
+let identifyData: {
+  userId: string;
+  deviceId: string;
+  traits: Record<string, any>;
+} | null = null;
+
+function getClient(): RudderAnalytics {
+  if (client) {
+    return client;
+  }
```
should wrap all these methods and global state in an Analytics class like in XDL https://github.com/expo/expo-cli/blob/main/packages/xdl/src/Analytics.ts

e.g. create an internal analytics class here, then export two versions of it with different write keys and have the caller specify which client they need

```
class AnalyticsClient {
  private readonly writeKey: string;
  private client?: RudderAnalytics;

  constructor(writeKey: string) {
    this.writeKey = writeKey;
  }

  function getClient(): RudderAnalytics {
    if (client) {
      return client;
    }

    client = new RudderAnalytics(
      this.writeKey,
      'https://cdp.expo.dev/v1/batch',
      {
        flushInterval: 300,
      }
    );
  
    // Install flush on exit...
    process.on('SIGINT', () => client?.flush?.());
    process.on('SIGTERM', () => client?.flush?.());
  
    return client;
  }
  ....
}

const xdlClient = new AnalyticsClient('1wHTzmVgmZvNjCalKL45chlc2VN');
const unifiedClient = new AnalyticsClient('1wabJGd5IiuF9Q8SGlcI90v8WTs');

export { xdlClient, unifiedClient };
```

when we need two clients this pattern keeps a better separation of state than having them all share the same global state. If you strongly want to retain the pattern you have here, you could export separate `logXDLEvent` & `logUnifiedEvent` methods, then have separate initialize and identify functions but i think it would b easier for errors to creep in (in both usage and development of analytics) that way.

EvanBacon committed:
Merge branch 'main' into @evanbacon/expo/cli/auth-commands

EvanBacon committed:
Fix up code

Update registerAsync.ts
Remove unused settings
Drop unused
Update otp.ts
Update whoamiAsync.ts

EvanBacon committed:
Update api-test.ts

kgc00 commented on packages/expo/cli/utils/analytics/rudderstackClient.ts:
```diff
@@ -0,0 +1,99 @@
+    return client;
+  }
+
+  client = new RudderAnalytics(
+    EXPO_STAGING || EXPO_LOCAL ? '1wpX20Da4ltFGSXbPFYUL00Chb7' : '1wpXLFxmujq86etH6G6cc90hPcC',
+    'https://cdp.expo.dev/v1/batch',
+    {
+      flushInterval: 300,
+    }
+  );
```
i added in a staging route for expo-cli events in our backend for this. Now that we are changing up the pipeline it seemed like a good time to do so.

kgc00 commented on packages/expo/cli/utils/analytics/rudderstackClient.ts:
```diff
@@ -0,0 +1,99 @@
+
+  ensureIdentified();
+}
+
+export function logEvent(name: string, properties: Record<string, any> = {}): void {
```
```suggestion

export enum AnalyticsEvent {
  ACTION = 'action', // generic event type which is used to determine the 'daily active user' stat, include an `action: expo ${subcommand}` property inside of the event properties object
}

export function logEvent(event: AnalyticsEvent, properties: Record<string, any> = {}): void {
```

EvanBacon replied:
I chose a string union because it's lighter weight when transpiled down to JS. This code won't ship with types and users shouldn't import the code directly, so there's no need to support environments that don't have strong autocomplete (i.e. environments where enums would make dx nicer).

kgc00 commented on packages/expo/cli/utils/analytics/rudderstackClient.ts:
```diff
@@ -0,0 +1,99 @@
+
+  const identity = { userId: userId ?? undefined, anonymousId: deviceId ?? uuidv4() };
+  getClient().track({
+    event: name,
```
```suggestion
    event,
```

EvanBacon committed:
Apply suggestions from code review

Co-authored-by: kgc00 <kirby@expo.io>

EvanBacon committed:
Update rudderstackClient.ts

EvanBacon committed:
Update rudderstackClient.ts
